### DISCUSSION TITLE
Implemented: Patch File Support for Splitting 1 Scan with multiple docs into multiple files

### DISCUSSION THREAD

**Discussion by smseidl** [paperless-ngx/paperless-ngx#285] [Feature Requests] [answered]
I would love to see Paperless-NGX add support for PATCH Pages to separate documents.  To make use of these today I have to NAPS2 scanning software on my PC to scan, split, and upload into my NAS Consumer directory.   It would be great if Paperless-NGX could handle this directly and split one scan into multiple documents.

http://www.alliancegroup.co.uk/patch-codes.htm
https://www.naps2.com/

**Comment by gador** [comment]
Ok, I looked into it a bit:

### Conditions

- pdf files containing multiple documents and `PATCHT` seperator barcode pages
- Adding these documents through email, web-ui and consumer folder

### How to separate the pages

1. Use an external program (like @smseidl )
2. Integrate, using a [pre-consumption script](https://paperless-ngx.readthedocs.io/en/latest/advanced_usage.html#pre-consumption-script)
3. Integrate, using paperless 

### Solution one

- Pros
  - Users are free to use any program they are familiar with
  - No changes to paperless is necessary
  - Can be used right now

- Cons
  - Can be cumbersome to use
  - Users need to take care of this step themselves
  - Reliance on third party applications
  - Availability of barcode/separator page recognition software (is there any other than naps2?)

### Solution two

- Pros
  - No big changes to paperless necessary

- Cons
  - Paperless (container or bare metal installation) will need the added python dependencies of the library used for the barcode recognition
  - Since the only parameter passed to the script is the filepath, it can only modify this file. So the other pages either get deleted (which isn't useful at all), or need to be saved to (e.g.) the consumption folder to be processed later.


### Solution three

- Pros
  - Every user profits of the feature, without manual user intervention (except update paperless)
  - it is integrated and the user won't need to meddle with additional scripts or programs
  - (unit)tests can be added
 
- Cons
  - More work to integrate properly (including tests)
  - Need some development time until its ready to use
  - Additional (python) dependencies

For solution two and three I already took a look at [pyzbar](https://pypi.org/project/pyzbar/) and [Pypdf2](https://pypi.org/project/PyPDF2/), which seem to do the trick.

For solution three, there are several possible ways to integrate with paperless:

1. Add to https://github.com/paperless-ngx/paperless-ngx/blob/e514b99c57f71476e7115468021254d0728d3535/src/documents/consumer.py#L166 
  This is the function which does the actual work of consuming the file. 
2. Add to the individual consumption processes like https://github.com/paperless-ngx/paperless-ngx/blob/e514b99c57f71476e7115468021254d0728d3535/src/documents/management/commands/document_consumer.py#L43
3. Add centrally to the task queue: https://github.com/paperless-ngx/paperless-ngx/blob/e514b99c57f71476e7115468021254d0728d3535/src/documents/tasks.py#L60


Solution one has the advantage of being centrally involved in the process. If adding a barcode recognition function call here, it would need to call itself `try_consume_file` again and again until all different documents within the file are processed. 
Solution two wouldn't need to do that, since it could call the function every time with the already split documents. It would need to be added to the mail fetcher and the web-ui task as well.
Solution three is even more central than solution one, since this is the task queue which all the fetchers (directory, email, web-ui) call.

I personally don't like solution one and two. Solution one could potentially end in a recursion if an error happens during barcode recognition. Solution two adds basically the same code at different source files, or would need to add a separate class. Even with a separate class, it would add the same functions calls at three different places.
Right now I think solution three seems viable.

Adding a barcode recognition routine centrally in the task queue would mean that code only gets added once and the function `try_consume_file` would get called with a unique document every time. 

I am still quite unfamiliar with the code and would greatly appreciate feedback and input from you guys, especially the @paperless-ngx/backend team, before I start implementing this feature. 

Thanks guys

**Reply by smseidl** [reply]
Wow - That's a lot of research.  I will try to review them to see if I can see anything.  I would be happy with Solution 2 (pre-consumption) but hadn't figured out a good way to dive into.  I think Solution 3 would be the best for all.

**Comment by gador** [comment]
So I did a bit more:

I decided to implement solution three. I added the code to tasks.py, since it is the central point for consuming. Feature now include:

- Add an option to enable or disable the barcode splitter. 
- Make the barcode string configurable
- If enabled, documents added through the UI, consume directory or email will be split upon the barcode pages
- the new files will pe prefixed with a count and copied to the consumption dir
- the original file will be deleted, once completely split
- the new files get consumed by the default consumer, since the barcode pages have been removed

Things I did:
- Added the logic
- Added the new package dependencies to the github ci and Dockerimage
- Added unittests for the new functions
- Added the config options
- Added a paragraph to the settings documentation
- Tested with sample pdfs from the web-ui using docker-compose on the local paperless-ngx instance

Things that still need to be done:
- Add to changelog
- Add to documentation / feature list
- **Way** more extensive testing. Including email and consumption directory
- Due to the library used, this will only work with pdfs right now. A TIFF Implementation is planed, though.

The code can be found in the branch [feature-barcode-splitter](https://github.com/paperless-ngx/paperless-ngx/tree/feature-barcode-splitter)

@paperless-ngx/backend 
@paperless-ngx/test

**Comment by gador** [comment, marked as answer]
I added a PR for easier code review (#532)

**Comment by smseidl** [comment]
@gador almost ready to do some testing on this... taken time to figure out how to get a second instance up and running on my RPI 4.  One question that I just came up with is on the type of bar codes it will split on.  Is it any bar code, or does it have to be a specific type (i.e. Patch T)?

**Reply by gador** [reply]
You can specify the content of the Barcode. Right now, it is the content of the patch-t Barcodes.
You can adjust it to the content of other Barcodes with
PAPERLESS_CONSUMER_BAROCDE_STRING environment variable.

Theoretically, you could use any Barcode pyzbar understands.

**Comment by smseidl** [comment]
@shamoon - not sure who to ping - but based on what is in the `beta` branch, I think this feature request should be closed... and I can't close it even though I opened it.

**Reply by shamoon** [reply]
Thanks! You cant "close" these but I'll lock it so people know its done, the PR is chosen as the "answer" too
